### ISSUE TITLE
[43] GLaM: Efficient Scaling of Language Models with Mixture-of-Experts

### ISSUE THREAD

**Issue by dhkim0225** [dhkim0225/1day_1paper#71]
[paper](https://arxiv.org/pdf/2112.06905.pdf)
[blog](https://ai.googleblog.com/2021/12/more-efficient-in-context-learning-with.html)

# Abstract & Intro
구글에서 새롭게 공개한 MoE 모델.
GPT-3 너무 크고 학습시간 길다. 효율적인 거 만들자.
- 1.2 trillion param (7x larger than GPT-3)
- 1/3 energy usage
- inference 시에도 절반의 computation 이면 됨.
![image](https://user-images.githubusercontent.com/16400591/146313187-4a5a369f-4fe4-47a7-bb43-b4ced8e1cdf3.png)

# Training Sets
Fair comparison 을 나름 신경 쓴 데이터 콜렉팅.
GPT-3 에서 collecting 한 방식 그대로 사용.
![image](https://user-images.githubusercontent.com/16400591/146314599-74ff7154-130c-467e-ae48-d5dc477cbbee.png)

~누군가 benchmark 용으로 데이터 collecting 방식도 통일 시켜줄 필요가 있을 듯.~
~성능 개선이 데이터에 의한 건지, 모델에 의한 건지 알 수 있는 방법이 없음.~

# Model Architecture
![image](https://user-images.githubusercontent.com/16400591/146353452-0efec0e6-15bd-4f0b-bb72-e161bfa76d04.png)

[GShard](https://arxiv.org/pdf/2006.16668.pdf) 부터 구글은 MoE 를 주목해 왔음.
당시에는 커다란 모델로 꽤나 관심을 끌었던 논문.
보통 transformer 를 scaling 한다고 하면, token input sequence 를 늘려서 attention 을 키우던가, layer 를 더 쌓았지, feed-forward 는 딱히 손 안댔었음. 
Gshard 모델은 당시에 MoE 를 이용해 feed-forward 를 엄청나게 불렸고, 꽤 잘 동작.

새롭게 제안하는 GLaM (**G**eneralist **La**nguage **M**odel) 에서 비슷하게 시도함.
gating function 을 이용해서 MoE 중 일부만을 사용. (딱 2개만 씀)
최종 output 은 2개의 expert output 을 weighted sum 해서 사용함.
![model_gif](https://blogger.googleusercontent.com/img/a/AVvXsEhKmA36iEaxIj0bSi7IUfctim-muB1B695evFlPNoqkiIbD_mVkkbdslk6HNwEXCBJmGsvxyDYsNTS7-YzZlblcQa2St498_RPwm72uz6jrK8mooPA6-wN4vqNswF-MG8JEPsgj7R4Rsm60Xh1xVLQtSeYrfhNEG4pr-7MbTHYdxjbug845Fy5smz17ZA=w402-h640)

그림에서 볼 수 있듯이, MoE 모델 이후에 non-MoE 모델이 하나 붙음

추가로 적용한건,
- per layer relative PE 사용 ~(gopher랑 같음)~
- non-MoE 모델의 feedforward 에 GLU (Gated Linear Unit) 사용
- LayerNorm 대신 RMSNorm 사용 ~(gopher랑 같음)~

[GSPMD](https://arxiv.org/pdf/2105.04663.pdf) 에서 제안했던 2d sharding 알고리즘으로 multi-node training 을 한다고 함.

## Model Family
![image](https://user-images.githubusercontent.com/16400591/146314783-048f291c-30a4-4786-bbeb-28e765164270.png)
![image](https://user-images.githubusercontent.com/16400591/146315062-f8646cc1-d207-4a70-ae22-930bf580db52.png)

## Training Details
130M 부터 1.2 trillion 까지 실험해 봤다고 한다
- max_seq_length 는 1024
- 각 batch input 은 1 million token
- dropout rate 는 0
- Adafactor (β1 = 0,  β2 = 0.99)
- LR ==> 1 - (step)^{-0.8} decay
- 첫 10K 동안은 LR 0.01 유지
- GShard 논문처럼 MoE auxiliary loss 추가
- SentencePiece (vocab size 256K)
- float32 weight + bfloat16 activation

## Training Tips
1. 더 작은 규모의 모델을 먼저 학습시킴. dataset이나 인프라의 잠재적인 문제를 가능한 한 빨리 알아챔.
2. gradient nan 뜨면 그냥 넘김 (inf 는 언제든 발생 가능)
3. 가끔 학습 망할 때는, 안정적이었던 곳에서부터 model resume. (ㄷㄷ...) 해당 시드 망했으니까, 랜덤 시드 다르게 해서 돌렸다 함. ~(reproducibility 개나 줘버려?)~

파라미터 튜닝 못할 정도로 모델이 커서 대충 조정한 걸로 통일해서 학습시켰다고 함

# Results
## vs GPT-3
역시나 비교군은 GPT-3
![image](https://user-images.githubusercontent.com/16400591/146350654-0446054c-6a9d-4697-99e6-552024d3b965.png)
![image](https://user-images.githubusercontent.com/16400591/146350720-dc7b7336-6f98-4682-8d62-6f5e9184d7db.png)

## TriviaQA
TriviaQA 를 통해 MoE 의 우월성 강조
![image](https://user-images.githubusercontent.com/16400591/146350840-18347717-0e32-4624-b17f-4f054a2136e9.png)
![image](https://user-images.githubusercontent.com/16400591/146350927-890a6df0-ac92-4547-a3de-5dfcca005338.png)

## ANLI
![image](https://user-images.githubusercontent.com/16400591/146350978-058d6088-5e71-4a04-9ff2-20c5a9cf47d1.png)

## Data-Filtering
GPT-3 의 filtering 전략은 훌륭했다.
![image](https://user-images.githubusercontent.com/16400591/146351048-11082871-7335-4125-925b-cf8cf705917c.png)

더 좋은 filtering 이 있다면 더 성능이 오를 것 같다고 말함.
~(gopher 에서 이미 더 좋은 filtering 을 만들었음 ㅋㅋ)~

## Scaling MoE
![image](https://user-images.githubusercontent.com/16400591/146351675-c4cf7a6e-163e-4ca9-a960-0e3f276a9601.png)

### 8 NLG task average
![image](https://user-images.githubusercontent.com/16400591/146351444-7f0e3d52-5c8d-4cff-bc20-941e06cfb113.png)

### 21 NLU task average
![image](https://user-images.githubusercontent.com/16400591/146351514-bce1657a-06a0-46ce-899d-6794b6972123.png)

# Representation and Society
생략.
